### PR TITLE
Move chat logic into separate service

### PR DESCRIPTION
**Pull request by dgieselaar**
Moves chat logic into a separate service. This allows plugins to register components conditionally, e.g. based on a has-data check. Additionally, adds a `get_apm_timeseries` function. The system message is now also constructed from registered context definitions.

### CODE REVIEW

File: x-pack/plugins/observability_ai_assistant/public/components/chat/chat_flyout.tsx [hunk 1]
```diff
@@ -47,14 +50,14 @@ export function ChatFlyout({
...
 
-  const service = useObservabilityAIAssistant();
-
   const { euiTheme } = useEuiTheme();
 
   const router = useObservabilityAIAssistantRouter();
 
   const knowledgeBase = useKnowledgeBase();
 
+  useEffect(() => {}, [knowledgeBase.status.value?.ready]);
```
Comment (CoenWarmer) [inline x-pack/plugins/observability_ai_assistant/public/components/chat/chat_flyout.tsx:59]: What does this do?

File: x-pack/plugins/observability_ai_assistant/public/functions/index.ts [hunk 2]
```diff
@@ -5,31 +5,68 @@
...
-  registerSummarisationFunction({ service, registerFunction });
-  registerRecallFunction({ service, registerFunction });
+      let description = `You have the ability to call Elasticsearch APIs or create visualisations using Lens.`;
+
+      if (isReady) {
+        description += `You can use the "summarise" functions to store new information you have learned in a knowledge database. Once you have established that you did not know the answer to a question, and the user gave you this information, it's important that you create a summarisation of what you have learned and store it in the knowledge database. When you create this summarisation, make sure you craft it in a way that can be recalled with a semantic search later.
+
+        Additionally, you can use the "recall" function to retrieve relevant information from the knowledge database. Using the "recall" function will allow you to create a much more tailored and improved user experience. Assume the user has told you before.
+      
+        For every question that the user asks, you, as the assistant, must query the knowledge base, no matter how confident you are in your response.
```
Comment (CoenWarmer) [inline x-pack/plugins/observability_ai_assistant/public/functions/index.ts:43]: Will this mean the user will, for every prompt, see a "the assistant has suggested 'recall'" and "you executed function 'recall'"?

File: x-pack/plugins/observability_ai_assistant/public/utils/get_timeline_items_from_conversation.tsx [hunk 3]
```diff
@@ -88,15 +84,13 @@ export function getTimelineItemsfromConversation({
               arguments: JSON.parse(functionCall.arguments || '{}'),
             });
 
-            const fn = functions.find((func) => func.options.name === message.message.name);
-
-            element = fn?.render ? (
+            element = (
```
Comment (CoenWarmer) [inline x-pack/plugins/observability_ai_assistant/public/utils/get_timeline_items_from_conversation.tsx:87]: How do you test if this element actually returns something?

**Review by CoenWarmer — COMMENTED** [review]
I suspect the current structure of Message can't capture all ChatItem states (executed a function, got a response, got a visualization or any combination thereof). I have done my best to fix this over the past days but you're rewinding stuff so I guess you have a better idea.

File: x-pack/plugins/observability_ai_assistant/public/components/chat/chat_flyout.tsx (hunk 1 above)
Comment (dgieselaar) [reply x-pack/plugins/observability_ai_assistant/public/components/chat/chat_flyout.tsx:59]: ah I need to re-create the chatService when this happens, will fix.

File: x-pack/plugins/observability_ai_assistant/public/functions/index.ts (hunk 2 above)
Comment (dgieselaar) [reply x-pack/plugins/observability_ai_assistant/public/functions/index.ts:43]: I'm not sure if I follow, this is part of the system prompt (that the user doesn't see).

File: x-pack/plugins/observability_ai_assistant/public/utils/get_timeline_items_from_conversation.tsx (hunk 3 above)
Comment (dgieselaar) [reply x-pack/plugins/observability_ai_assistant/public/utils/get_timeline_items_from_conversation.tsx:87]: There's no difference from what it was before - the RenderFunction component returns the rendered element, or `null`, which has the same end result as before. It's just encapsulated in the RenderFunction now.

Comment (dgieselaar) [reply x-pack/plugins/observability_ai_assistant/public/utils/get_timeline_items_from_conversation.tsx:87]: Actually, this isn't true, there's no difference in rendering, but there is a difference in the value of `element`. Let me check the ramifications.

Comment (dgieselaar) [reply x-pack/plugins/observability_ai_assistant/public/utils/get_timeline_items_from_conversation.tsx:87]: fixed in https://github.com/dgieselaar/kibana/pull/38/commits

File: x-pack/plugins/observability_ai_assistant/public/components/chat/chat_flyout.tsx (hunk 1 above)
Comment (dgieselaar) [reply x-pack/plugins/observability_ai_assistant/public/components/chat/chat_flyout.tsx:59]: I've removed the hook here [`a7ccdf1` (#38)](https://github.com/dgieselaar/kibana/pull/38/commits/a7ccdf1ad9b47d1fc1c1eb7c97b7c0b41f6da181), the refresh logic needs to happen in a separate place.

**Pull request merged**